packages/agent/src/openid4vc/deferredCredentialRecord.ts

```diff
@@ -0,0 +1,119 @@
+
+  try {
+    let accessToken: OpenId4VciRequestTokenResponse = {
+      ...deferredCredential.accessToken,
+      dpop: deferredCredential.accessToken.dpop
+        ? {
+            ...deferredCredential.accessToken.dpop,
+            jwk: Kms.PublicJwk.fromUnknown(deferredCredential.accessToken.dpop.jwk),
+          }
+        : undefined,
+    }
+
+    // Determine whether or not to use the refresh token to get a new access token.
+    // In the future, this could be improved with a retry: first try the token
+    // we already have, and if that fails, refresh the token and try again.
+    if (accessToken.refreshToken) {
+      console.log('Going to Refresh access token')
+      accessToken = await agent.modules.openid4vc.holder.refreshToken({
+        issuerMetadata,
+        clientId,
+        refreshToken: accessToken.refreshToken,
+        authorizationServer: accessToken.authorizationServer,
+        dpop: accessToken.dpop,
+      })
+    }
+
+    // Fetch the credentials from the deferred credential endpoint
+    const { credentials, deferredCredentials } = await receiveDeferredCredentialFromOpenId4VciOffer({
+      accessToken,
+      issuerMetadata,
+      agent,
+      deferredCredentialResponse,
+    })
+
+    if (deferredCredentials.length && credentials.length) {
+      throw new Error('Received both immediate and deferred credentials in OpenID4VCI response')
+    }
+
+    if (credentials.length) {
+      for (const { credential } of credentials) {
+        await storeCredential(agent, credential)
+      }
+
+      const { issuer: issuerDisplay } = getCredentialDisplayWithDefaults(
+        getOpenId4VcCredentialDisplay(
+          extractOpenId4VcCredentialMetadata(credentials[0].credentialConfiguration, {
+            display: issuerMetadata.credentialIssuer?.display,
+            id: issuerMetadata.credentialIssuer?.credential_issuer,
+          })
+        )
+      )
+
+      await storeReceivedActivity(agent, {
+        // FIXME: Should probably be the `iss`, but then we can't show it before we retrieved
+        // the credential. Signed issuer metadata is the solution.
+        entityId: issuerMetadata?.credentialIssuer.credential_issuer,
+        host: issuerDisplay.domain,
+        name: issuerDisplay.name,
+        logo: issuerDisplay.logo,
+        backgroundColor: '#ffffff', // Default to a white background for now
+        deferredCredentials: [],
+        credentialIds: credentials.map(({ credential }) => getCredentialForDisplayId(credential)),
+      })
+
+      await deleteDeferredCredential(agent, deferredCredential.id)
+    } else {
+      await updateDeferredCredential(agent, {
+        id: deferredCredential.id,
+        lastCheckedAt: new Date().toISOString(),
+        // Update access token with a new one if we got a new refresh token.
+        // Otherwise, keep the existing one.
+        accessToken: accessToken.refreshToken
```

Comment:
The access token response returns `expires_in`. Should we somehow store this on the record to know:
- when the deferred issuance failed due to the token having expired (in case no refresh token)
- whether we should retrieve a new access token or use the existing one (in case there's a refresh token(

Reply:
I updated the fetching logic to only get a refresh token if the access token is indeed expired.